#  'Assembly Accession',
#  'ANI Best ANI match ANI',
#  'ANI Best ANI match Assembly',
#  'ANI Best ANI match Assembly Coverage',
#  'ANI Best ANI match Type Category',
#  'ANI Best ANI match Organism',
#  'ANI Best ANI match Type Assembly Coverage',
#  'ANI Best match status',
#  'ANI Category',
#  'ANI Check status',
#  'ANI Comment',
#  'ANI Declared ANI match ANI',
#  'ANI Declared ANI match Assembly',
#  'ANI Declared ANI match Assembly Coverage',
#  'ANI Declared ANI match Type Category',
#  'ANI Declared ANI match Organism',
#  'ANI Declared ANI match Type Assembly Coverage',
#  'ANI Submitted organism',
#  'ANI Submitted species',
#  'Annotation BUSCO Complete ',
#  'Annotation BUSCO Duplicated ',
#  'Annotation BUSCO Fragmented ',
#  'Annotation BUSCO Lineage ',
#  'Annotation BUSCO Missing ',
#  'Annotation BUSCO Single Copy ',
#  'Annotation BUSCO Total Count ',
#  'Annotation BUSCO Version ',
#  'Annotation Count Gene Non-coding',
#  'Annotation Count Gene Other',
#  'Annotation Count Gene Protein-coding',
#  'Annotation Count Gene Pseudogene',
#  'Annotation Count Gene Total',
#  'Annotation Method',
#  'Annotation Name',
#  'Annotation Pipeline',
#  'Annotation Provider',
#  'Annotation Release Date',
#  'Annotation Release Version',
#  'Annotation Report URL',
#  'Annotation Software Version',
#  'Annotation Status',
#  'Assembly Assembly Method',
#  'Assembly Atypical Is Atypical',
#  'Assembly Atypical Warnings',
#  'Assembly BioProject Accession',
#  'Assembly BioProject Lineage Accession',
#  'Assembly BioProject Lineage Parent Accession',
#  'Assembly BioProject Lineage Parent Accessions',
#  'Assembly BioProject Lineage Title',
#  'Assembly BioSample Accession',
#  'Assembly BioSample Age ',
#  'Assembly BioSample Attribute Name',
#  'Assembly BioSample Attribute Value',
#  'Assembly BioSample Biomaterial provider ',
#  'Assembly BioSample BioProject Accession',
#  'Assembly BioSample BioProject Parent Accession',
#  'Assembly BioSample BioProject Parent Accessions',
#  'Assembly BioSample BioProject Title',
#  'Assembly BioSample Breed ',
#  'Assembly BioSample Collected by ',
#  'Assembly BioSample Collection date ',
#  'Assembly BioSample Cultivar ',
#  'Assembly BioSample Description Comment',
#  'Assembly BioSample Description Organism Common Name',
#  'Assembly BioSample Description Organism Infraspecific Names Breed',
#  'Assembly BioSample Description Organism Infraspecific Names Cultivar',
#  'Assembly BioSample Description Organism Infraspecific Names Ecotype',
#  'Assembly BioSample Description Organism Infraspecific Names Isolate',
#  'Assembly BioSample Description Organism Infraspecific Names Sex',
#  'Assembly BioSample Description Organism Infraspecific Names Strain',
#  'Assembly BioSample Description Organism Name',
#  'Assembly BioSample Description Organism Pangolin Classification',
#  'Assembly BioSample Description Organism Taxonomic ID',
#  'Assembly BioSample Description Title',
#  'Assembly BioSample Development stage ',
#  'Assembly BioSample Ecotype ',
#  'Assembly BioSample Geographic location ',
#  'Assembly BioSample Host ',
#  'Assembly BioSample Host disease ',
#  'Assembly BioSample Identified by ',
#  'Assembly BioSample Sample Identifiers Database',
#  'Assembly BioSample Sample Identifiers Label',
#  'Assembly BioSample Sample Identifiers Value',
#  'Assembly BioSample IFSAC category ',
#  'Assembly BioSample Isolate ',
#  'Assembly BioSample Isolate name alias ',
#  'Assembly BioSample Isolation source ',
#  'Assembly BioSample Last updated',
#  'Assembly BioSample Latitude / Longitude ',
#  'Assembly BioSample Models',
#  'Assembly BioSample Owner Contact Lab',
#  'Assembly BioSample Owner Name',
#  'Assembly BioSample Package',
#  'Assembly BioSample Project name  ',
#  'Assembly BioSample Publication date',
#  'Assembly BioSample Sample name ',
#  'Assembly BioSample Serotype ',
#  'Assembly BioSample Serovar ',
#  'Assembly BioSample Sex ',
#  'Assembly BioSample Source type ',
#  'Assembly BioSample Status Status',
#  'Assembly BioSample Status When',
#  'Assembly BioSample Strain ',
#  'Assembly BioSample Sub-species ',
#  'Assembly BioSample Submission date',
#  'Assembly BioSample Tissue ',
#  'Assembly Blast URL',
#  'Assembly Description',
#  'Assembly Level',
#  'Assembly Linked Assembly Accession',
#  'Assembly Linked Assembly Type',
#  'Assembly Name',
#  'Assembly Notes ',
#  'Assembly Paired Assembly Accession',
#  'Assembly Paired Assembly Changed',
#  'Assembly Paired Assembly Manual Diff',
#  'Assembly Paired Assembly Name',
#  'Assembly Paired Assembly Only Genbank',
#  'Assembly Paired Assembly Only RefSeq',
#  'Assembly Paired Assembly Status',
#  'Assembly Refseq Category',
#  'Assembly Release Date',
#  'Assembly Sequencing Tech',
#  'Assembly Status',
#  'Assembly Submitter',
#  'Assembly Suppression Reason',
#  'Assembly Synonym',
#  'Assembly Type',
#  'Assembly Stats Contig L50',
#  'Assembly Stats Contig N50',
#  'Assembly Stats Gaps Between Scaffolds Count',
#  'Assembly Stats GC Count',
#  'Assembly Stats GC Percent',
#  'Assembly Stats Genome Coverage',
#  'Assembly Stats Number of Component Sequences',
#  'Assembly Stats Number of Contigs',
#  'Assembly Stats Number of Organelles',
#  'Assembly Stats Number of Scaffolds',
#  'Assembly Stats Scaffold L50',
#  'Assembly Stats Scaffold N50',
#  'Assembly Stats Total Number of Chromosomes',
#  'Assembly Stats Total Sequence Length',
#  'Assembly Stats Total Ungapped Length',
#  'CheckM completeness',
#  'CheckM completeness percentile',
#  'CheckM contamination',
#  'CheckM marker set',
#  'CheckM marker set rank',
#  'CheckM species tax id',
#  'CheckM version',
#  'Current Accession',
#  'Organelle Assembly Name',
#  'Organelle BioProject Accessions',
#  'Organelle Description',
#  'Organelle Infraspecific Name',
#  'Organelle Submitter',
#  'Organelle Total Seq Length',
#  'Organism Common Name',
#  'Organism Infraspecific Names Breed',
#  'Organism Infraspecific Names Cultivar',
#  'Organism Infraspecific Names Ecotype',
#  'Organism Infraspecific Names Isolate',
#  'Organism Infraspecific Names Sex',
#  'Organism Infraspecific Names Strain',
#  'Organism Name',
#  'Organism Pangolin Classification',
#  'Organism Taxonomic ID',
#  'Source Database',
#  'Type Material Display Text',
#  'Type Material Label',
#  'WGS contigs URL',
#  'WGS project accession',
#  'WGS URL'

# %%
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import pathlib

fig_fld = pathlib.Path("figs/f01")
fig_fld.mkdir(parents=True, exist_ok=True)

fname = "../data/dset_info.tsv"
df = pd.read_csv(
    fname,
    sep="\t",
    dtype={_: str for _ in [53, 79, 83, 85, 93, 97, 99, 103, 120, 168, 169]},
)

# %%

lab = "Assembly BioSample Collection date "
df[lab] = pd.to_datetime(df[lab], errors="coerce")

fig, ax = plt.subplots(figsize=(8, 4))
sns.histplot(data=df, x=lab, bins=30, ax=ax)
sns.despine()
plt.tight_layout()
plt.savefig(fig_fld / "assembly_release_date_hist.png")
plt.show()

# %%
lab = "Assembly BioSample Geographic location "
nlab = "Geographic Location"
df[nlab] = df[lab].str.split(":").str[0].str.strip()

mp = {
    "unknown": None,
    "Unknown": None,
    "not collected": None,
    "not applicable": None,
    "Not Applicable": None,
    "not determined": None,
    "missing": None,
    "Missing": None,
    "-": None,
}
df[nlab] = df[nlab].replace(mp)

# sort by count
df[nlab] = df[nlab].astype("category")
cat_ord = df[nlab].value_counts().sort_values(ascending=False).index
df[nlab].cat.set_categories(cat_ord, ordered=True)
fig, ax = plt.subplots(figsize=(5, 10))
sns.countplot(data=df, y=nlab, ax=ax, order=cat_ord)
ax.set_xscale("log")
sns.despine()
plt.tight_layout()
plt.savefig(fig_fld / "assembly_geo_location_count.png")
plt.show()
# %%

sns.histplot(data=df, x="Assembly Stats Total Sequence Length", bins=30)
# %%
